lighthouse-core/audits/unused-css-rules.js

```diff
@@ -0,0 +1,154 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit');
+const Formatter = require('../formatters/formatter');
+
+const PREVIEW_LENGTH = 100;
+const ALLOWABLE_UNUSED_RULES_RATIO = 0.10;
+
+class UnusedCSSRules extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Best Practices',
+      name: 'unused-css-rules',
+      description: 'Site does not have more than 10% unused CSS',
+      helpText: 'Remove unused rules from stylesheets to reduce unnecessary ' +
+          'bytes consumed by network activity. [Learn more](https://developers.google.com/speed/docs/insights/OptimizeCSSDelivery)',
+      requiredArtifacts: ['CSSUsage', 'Styles']
+    };
+  }
+
+  /**
+   * @param {!Array.<Object>} styles The output of the Styles gatherer.
+   * @return {!Object} A map of styleSheetId to stylesheet information.
+   */
+  static indexStylesheetsById(styles) {
+    return styles.reduce((indexed, stylesheet) => {
+      indexed[stylesheet.header.styleSheetId] = Object.assign({
+        used: [],
+        unused: [],
+      }, stylesheet);
+      return indexed;
+    }, {});
+  }
+
+  /**
+   * Counts the number of unused rules and adds count information to sheets.
+   * @param {!Array.<RuleUsage>} rules The output of the CSSUsage gatherer.
+   * @param {!Object} indexedStylesheets Stylesheet information indexed by id.
+   * @return {number} The number of unused rules.
+   */
+  static countUnusedRules(rules, indexedStylesheets) {
+    let unused = 0;
+
+    rules.forEach(rule => {
+      const stylesheetInfo = indexedStylesheets[rule.styleSheetId];
+
+      if (rule.used) {
+        stylesheetInfo.used.push(rule);
+      } else {
+        unused++;
+        stylesheetInfo.unused.push(rule);
+      }
+    });
+
+    return unused;
+  }
+
+  /**
+   * @param {Object} stylesheetInfo The stylesheetInfo object.
```

Comment:
this might be too ridiculous to do inline, so you may just stick to `@param {!Object}` (or we can make a typedef somewhere if we want vscode/ts/closure support), but would be `@param {{used: !Array, unused: !Array, content: string, header: {sourceURL: string}}}`